src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs

```diff
@@ -204,12 +202,22 @@ private SingleTypeDeclaration FirstDeclarationWithExplicitBases()
             return null;
         }
 
-        internal Tuple<NamedTypeSymbol, ImmutableArray<NamedTypeSymbol>> GetDeclaredBases(ConsList<Symbol> basesBeingResolved)
+        internal Tuple<NamedTypeSymbol, ImmutableArray<NamedTypeSymbol>> GetDeclaredBases(ConsList<Symbol> basesBeingResolved, bool ignoreNonNullTypesAttribute)
         {
             if (ReferenceEquals(_lazyDeclaredBases, null))
             {
-                DiagnosticBag diagnostics = DiagnosticBag.GetInstance();
-                if (Interlocked.CompareExchange(ref _lazyDeclaredBases, MakeDeclaredBases(basesBeingResolved, diagnostics), null) == null)
+                DiagnosticBag diagnostics;
+                if (ignoreNonNullTypesAttribute)
+                {
+                    diagnostics = DiagnosticBag.GetInstance();
+                    Interlocked.CompareExchange(ref _lazyDeclaredBasesIgnoringNonNullTypesAttribute, MakeDeclaredBases(basesBeingResolved, ignoreNonNullTypesAttribute, diagnostics), null);
+                    diagnostics.Free();
+
+                    return _lazyDeclaredBasesIgnoringNonNullTypesAttribute;
+                }
+
+                diagnostics = DiagnosticBag.GetInstance();
+                if (Interlocked.CompareExchange(ref _lazyDeclaredBases, MakeDeclaredBases(basesBeingResolved, ignoreNonNullTypesAttribute, diagnostics), null) == null)
```

Comment:
>MakeDeclaredBases(basesBeingResolved, ignoreNonNullTypesAttribute, diagnostics), [](start = 72, length = 80)

Do we need to rebind the bases or is the result just a transformation of `_lazyDeclaredBasesIgnoringNonNullTypesAttribute`?
```
NonNullTypes ?
    _lazyDeclaredBasesIgnoringNonNullTypesAttribute :
    _lazyDeclaredBasesIgnoringNonNullTypesAttribute.
        SetUnknownNullabilityForReferenceTypes();
```

Reply:
That's a good idea, but I think we need to re-bind so that we get the diagnostics.

Reply:
We could report the diagnostics from calculating `_lazyDeclaredBasesIgnoringNonNullTypesAttribute`.

---
In reply to: [202067708](https://github.com/dotnet/roslyn/pull/28320#discussion_r202067708) [](ancestors = 202067708)

Reply:
If we hold onto the `DiagnosticBag` instance, we now have to worry about its lifetime (when does it need to be freed? maybe add a destructor?). Or we should just allocate one instead?

Reply:
We could report the diagnostics when `_lazyDeclaredBasesIgnoringNonNullTypesAttribute` is computed.

In short, `GetDeclaredBases()` would have two steps, binding without nullability (where diagnostics are reported), and transforming to include nullability. Similar to `PENamedTypeSymbol.GetDeclaredBaseType()`.

---
In reply to: [202075635](https://github.com/dotnet/roslyn/pull/28320#discussion_r202075635) [](ancestors = 202075635)

Reply:
I don't think that works. The code path that computes `_lazyDeclaredBasesIgnoringNonNullTypesAttribute` starts in a method that discards diagnostics (that's why there were no duplicate diagnostics, see your [earlier comment](https://github.com/dotnet/roslyn/pull/28320#discussion_r201404641)).

Reply:
Diagnostics would be reported indirectly, through the call to `AddDeclarationDiagnostics()`.

---
In reply to: [202080764](https://github.com/dotnet/roslyn/pull/28320#discussion_r202080764) [](ancestors = 202080764)

Reply:
Ran into one more regression, in the `ErrorTypeSymbolTests.ConstructedErrorTypes` symbols test. That test expects specific error types (`ExtendedErrorTypeSymbol`), but the use of `SetUnknownNullabilityForReferenceTypes` changes the types.

Also, while investigating that, I noticed that the `SetUnknownNullabilityForReferenceTypes` method only patches up `A<T!>.B<U!>` to `A<T!>.B<U>`. Is that what you expected?

I want to go back to the un-optimized approach that works. I'm happy to leave a PROTOTYPE comment to revisit.